packages/editor/src/store/reducer.js

```diff
@@ -352,63 +419,114 @@ export const editor = flow( [
 		byClientId( state = {}, action ) {
 			switch ( action.type ) {
 				case 'SETUP_EDITOR_STATE':
-					return getFlattenedBlocks( action.blocks );
+					return getFlattenedBlocksWithoutAttributes( action.blocks );
 
 				case 'RECEIVE_BLOCKS':
 					return {
 						...state,
-						...getFlattenedBlocks( action.blocks ),
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
 					};
 
-				case 'UPDATE_BLOCK_ATTRIBUTES':
+				case 'UPDATE_BLOCK':
 					// Ignore updates if block isn't known
 					if ( ! state[ action.clientId ] ) {
 						return state;
 					}
 
-					// Consider as updates only changed values
-					const nextAttributes = reduce( action.attributes, ( result, value, key ) => {
-						if ( value !== result[ key ] ) {
-							result = getMutateSafeObject( state[ action.clientId ].attributes, result );
-							result[ key ] = value;
-						}
-
-						return result;
-					}, state[ action.clientId ].attributes );
-
-					// Skip update if nothing has been changed. The reference will
-					// match the original block if `reduce` had no changed values.
-					if ( nextAttributes === state[ action.clientId ].attributes ) {
+					// Do nothing if only attributes change.
+					const changes = omit( action.updates, 'attributes' );
+					if ( isEmpty( changes ) ) {
 						return state;
 					}
 
-					// Otherwise merge attributes into state
 					return {
 						...state,
 						[ action.clientId ]: {
 							...state[ action.clientId ],
-							attributes: nextAttributes,
+							...changes,
 						},
 					};
 
+				case 'INSERT_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REPLACE_BLOCKS':
+					if ( ! action.blocks ) {
+						return state;
+					}
+
+					return {
+						...omit( state, action.clientIds ),
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REMOVE_BLOCKS':
+					return omit( state, action.clientIds );
+			}
+
+			return state;
+		},
+
+		attributes( state = {}, action ) {
+			switch ( action.type ) {
+				case 'SETUP_EDITOR_STATE':
+					return getFlattenedBlockAttributes( action.blocks );
+
+				case 'RECEIVE_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlockAttributes( action.blocks ),
+					};
+
 				case 'UPDATE_BLOCK':
-					// Ignore updates if block isn't known
-					if ( ! state[ action.clientId ] ) {
+					// Ignore updates if block isn't known or there are no attribute changes.
+					if ( ! state[ action.clientId ] || ! action.updates.attributes ) {
 						return state;
 					}
 
 					return {
 						...state,
 						[ action.clientId ]: {
 							...state[ action.clientId ],
-							...action.updates,
+							...action.updates.attributes,
 						},
 					};
 
+				case 'UPDATE_BLOCK_ATTRIBUTES':
```

Comment:
Kinda makes me wish we'd normalized `UPDATE_BLOCK_ATTRIBUTES` to use `UPDATE_BLOCK`, but it's also not something I think we ought to be addressing here.